GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,465 @@
+
+  !> Main entry point for client code.
+  public :: gx_minimax_grid
+
+contains
+
+  !> \brief Compute minimax grid for RPA energy and GW calculation on imaginary time/frequency domain.
+  !! @param[in] num_points: Number of mesh points.
+  !! @param[in] e_min: Minimum transition energy. Arbitrary units as we only need e_max/e_min
+  !! @param[in] e_max: Maximum transition energy.
+  !! @param[out] tau_points: imaginary time grid points
+  !! @param[out] tau_weights: weights for imaginary time grid points weights
+  !! @param[out] omega_points: imaginary frequency grid points
+  !! @param[out] omega_weights: weights for imaginary frequency grid points
+  !! @param[out] cosft_wt: weights for tau -> omega cosine transform. cos(w*t) factor is included.
+  !! @param[out] cosft_tw: weights for omega -> tau cosine transform. cos(w*t) factor is included.
+  !! @param[out] sinft_wt: weights for tau -> omega sine transform. sin(w*t) factor is included.
+  !! @param[out] max_errors: Max error for the three kind of transforms (same order as previous args)
+  !! @param[out] cosft_duality_error. Max_{ij} |AB - I| where A and B are the cosft_wt and cosft_tw matrices.
+  !! @param[out] ierr: Exit status
+  subroutine gx_minimax_grid(num_points, e_min, e_max, &
+       tau_points, tau_weights, omega_points, omega_weights, &
+       cosft_wt, cosft_tw, sinft_wt, &
+       max_errors, cosft_duality_error, ierr)
+
+    integer, intent(in)                               :: num_points
+    real(kind=dp), intent(in)                         :: e_min, e_max
+    real(kind=dp), allocatable, dimension(:), &  
+         intent(out)                                  :: tau_points, tau_weights
+    real(kind=dp), allocatable, dimension(:), &
+         intent(out)                                  :: omega_points(:), omega_weights(:)
+    real(kind=dp), allocatable, dimension(:, :), &
+         intent(out)                                  :: cosft_wt(:, :), cosft_tw(:, :), sinft_wt(:, :)
+    real(kind=dp), intent(out)                        :: max_errors(3), cosft_duality_error
+    integer, intent(out)                              :: ierr
+
+    ! Internal variables
+    integer, parameter                                :: cos_t_to_cos_w = 1
+    integer, parameter                                :: cos_w_to_cos_t = 2
+    integer, parameter                                :: sin_t_to_sin_w = 3
+    integer                                           :: i_point, j_point
+    real(kind=dp)                                     :: e_range, scaling
+    real(kind=dp), dimension(:), allocatable          :: x_tw
+    real(kind=dp), dimension(:, :), allocatable       :: mat
+
+    ! Begin work
+    e_range = e_max/e_min
+    ierr = 0
+
+    allocate (x_tw(2*num_points))
+
+    call get_points_weights_omega(num_points, e_range, x_tw, ierr)
+    if (ierr /= 0) return
+
+    allocate (omega_points(num_points))
+    allocate (omega_weights(num_points))
+
+    do i_point = 1, num_points
```

Comment:
A general point (and this is just a matter of taste so don't change it if you like it), but I can see i loops over points from the loop limit. Writing `i_point` is verbose and makes the expressions clunkier.  I would stick to i and j